keystore/keystore.go

```diff
@@ -0,0 +1,129 @@
+package keystore
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts/keystore"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/pborman/uuid"
+	"github.com/status-im/status-go/extkeys"
+	"github.com/status-im/status-go/keychain"
+)
+
+func NewKeystore(directory string, keychain keychain.Keychain) *Keystore {
+	return &Keystore{
+		directory: directory,
+		keychain:  keychain,
+	}
+}
+
+type Keystore struct {
+	directory string
+	keychain  keychain.Keychain
+}
+
+func (k *Keystore) ImportKey(ekey *extkeys.ExtendedKey, auth string) (common.Address, error) {
+	// TODO zero ecdsa key and marshalled bytes
+	key, err := newKeyForPurposeFromExtendedKey(extkeys.KeyPurposeWallet, ekey)
+	if err != nil {
+		return common.Address{}, err
+	}
+	path := filepath.Join(k.directory, key.Address.String())
+	if _, err := os.Stat(path); !os.IsNotExist(err) {
+		return common.Address{}, fmt.Errorf("account with address %x already exists", key.Address)
+	}
+	bytes, err := json.Marshal(key)
+	if err != nil {
+		return common.Address{}, err
+	}
+	err = k.keychain.CreateKey(auth)
+	if err != nil {
+		return common.Address{}, err
+	}
+	encrypted, err := k.keychain.Encrypt(auth, bytes)
+	if err != nil {
+		return common.Address{}, err
+	}
+	return key.Address, ioutil.WriteFile(path, encrypted, 0666)
+}
+
+func (k *Keystore) GetDecryptedKey(address common.Address, auth string) (*keystore.Key, error) {
+	// TODO zero decrypted bytes
+	path := filepath.Join(k.directory, address.String())
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	bytes, err := ioutil.ReadAll(f)
+	if err != nil {
+		return nil, err
+	}
+	keydata, err := k.keychain.Decrypt(auth, bytes)
+	if err != nil {
+		return nil, err
+	}
+	var key *keystore.Key
+	err = json.Unmarshal(keydata, key)
+	if err != nil {
+		return nil, err
+	}
+	return key, nil
+}
+
+func (k *Keystore) Accounts() (rst []common.Address, err error) {
+	files, err := ioutil.ReadDir(k.directory)
+	if err != nil {
+		return nil, err
+	}
+	for _, f := range files {
+		if f.IsDir() {
+			continue
+		}
+		parts := strings.Split(f.Name(), "/")
+		if len(parts) < 2 {
+			continue
+		}
+		rst = append(rst, common.HexToAddress(parts[len(parts[-1])]))
```

Comment:
Isn't this account file a JSON file we can get an address from? It looks odd we get this from a file name.

Reply:
that file will be encrypted with key from keychain and we will store only encrypted version on disk

Reply:
btw it is not fully implemented, i will check what additional info ethereum keystore keeps and will adjust keystore accordingly. i just needed something for an e2e test

Reply:
@dshulyak @adambabik  are we going to encrypt the full json file or do we want to use the normal keystore file encryption used by go-ethereum but using the keychain key instead of the normal password?

Reply:
we can encrypt only private key bytes. what is normal keystore file encryption? i think they encrypt private key using key derived from password and then put that in file with more metadata.

but this PR is on hold right now, more work is required with ios and android. not sure if i abandon it or not

Reply:
> i think they encrypt private key using key derived from password and then put that in file with more metadata.

yes exactly, I thought we were using the keychain key as password so that we didn't need to implement more things

Reply:
the idea was to encrypt ethereum keys with keychain key. using a key as a password to keystore is an interesting option that will simplify implementation for sure. do you know if it is possible to freely export key generated on keychain?

on android i see:

> Key material never enters the application process. When an application performs cryptographic operations using an Android Keystore key, behind the scenes plaintext, ciphertext, and messages to be signed or verified are fed to a system process which carries out the cryptographic operations. If the app's process is compromised, the attacker may be able to use the app's keys but cannot extract their key material (for example, to be used outside of the Android device).